### PR TITLE
Dockerize pipeline

### LINKED ISSUE #13: Dockerize pipeline
For creating a consistent environment for development and deployment.

### PR DESCRIPTION
**Pull request by bvtsang**
Before starting, make sure that `docker-compose` is installed.

To run:

```sh
$ docker-compose build
$ docker-compose up -d
$ docker-compose run -d app rake db:create
```

TODO:

- [x] Figure out how to boot everything up in one command
- [x] Check for redundant statements in Docker files
- [x] README: update prerequisites
- [x] README: update build instructions
- [x] README: add development workflow (Modifying files? Running commands (e.g. `rake`)?)

In the future, I may want to namespace `docker` commands in `rake` (e.g. `rake docker:blah`)

  - See https://github.com/new-xkit/pipeline/blob/master/lib/tasks/ci.rake

Closes #13

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+# Copy the Gemfile as well as the Gemfile.lock and install
+# the RubyGems. This is a separate step so the dependencies
+# will be cached unless changes to one of those two files
+# are made.
+COPY Gemfile Gemfile.lock ./
+RUN gem install bundler && bundle install --jobs 20 --retry 5
+
+# Copy the main application.
+COPY . ./
```
Comment (0xazure) [inline Dockerfile]: I don't think we want to `COPY` the app files, because if we did, any changes we'd make during development wouldn't be reflected on the server, right?  We'd have to restart the Docker image every time we wanted to see the changes.

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+db:
+  image: postgres:9.4
+  ports:
+    - "5432"
+app:
+  build: .
+  command: bundle exec rails server -p 3000 -b '0.0.0.0'
```
Comment (0xazure) [inline docker-compose.yml]: If we've already added a `CMD` in the Dockerfile, do we need to reiterate it here, or is `docker-compose` smart enough to figure it out?

**Comment by 0xazure** [comment]
> Figure out how to boot everything up in one command

We can probably modify the `bin/setup{,.bat}` script to handle this for us.  It should check if the requisite `docker*` commands are present on the host system and either print instructions on how to get set up or print a link to the documentation if they aren't found.  Otherwise it can start the Docker VM if it's not already running and start the Dockerfile.

File: docker-compose.yml (hunk 2 above)
Comment (bvtsang) [inline docker-compose.yml]: This [overrides the default command](https://docs.docker.com/compose/yml/#command) so we don't need the one in the Dockerfile. Nice catch.

File: app/assets/javascripts/application.js [hunk 3]
```diff
@@ -12,5 +12,4 @@
 //
 //= require jquery
 //= require jquery_ujs
-//= require turbolinks
```
Comment (0xazure) [inline app/assets/javascripts/application.js]: We missed this in the initial Rails setup, but it should probably be in a separate PR.

File: bin/setup [hunk 4]
```diff
@@ -5,25 +5,18 @@ require 'pathname'
...
-  system "bin/rake db:setup"
-
-  puts "\n== Removing old logs and tempfiles =="
-  system "rm -f log/*"
-  system "rm -rf tmp/cache"
-
-  puts "\n== Restarting application server =="
-  system "touch tmp/restart.txt"
+    puts "Initializing the database..."
+    system "docker-compose run -d app rake db:create"
```
Comment (0xazure) [inline bin/setup]: What's the difference between `rake db:create` and `rake db:setup`?

File: docker-compose.yml (hunk 2 above)
Comment (0xazure) [inline docker-compose.yml]: We probably want to keep the command in the Dockerfile as opposed to this one, no?  My understanding is that Dockerfiles can stand on their own (i.e. we could use it without `docker-compose`), but without the Dockerfile `CMD` we would lose that ability.

File: README.md [hunk 5]
```diff
@@ -1 +1,19 @@
...
+
+```sh
+ruby bin/setup
+```
+
+## Development Workflow
+
+1. Run Pipeline via the build instructions
+2. Make your code changes locally and they will be automatically reflected within the Docker container
+3. If you need to run commands (e.g. `rake`): `docker-compose run -d app rake db:create`
```
Comment (0xazure) [inline README.md]: I would re-word this as follows:

3. If you need to run commands (e.g. `rake`), prefix the command with `docker-compose run -d app` (e.x. `docker-compose run -d app rake db:create`).

File: bin/setup (hunk 4 above)
Comment (bvtsang) [inline bin/setup]: I've already answered in our Gitter, but sharing [the answer](http://jacopretorius.net/2014/02/all-rails-db-rake-tasks-and-what-they-do.html) here for posterity.

Quoting the relevant parts:

> **db:create** Creates the database for the current RAILS_ENV environment. If RAILS_ENV is not specified it defaults to the _development_ and _test_ databases.
> **db:setup** Runs **db:schema:load** and **db:seed**.
> **db:schema:load** Loads the schema into the current environment’s database.
> **db:seed** Runs the _db/seeds.rb_ file.

**Comment by 0xazure** [comment]
Given the current known issues with `bin/setup`, I'd like to propose reverting the changes to that file and simply documenting the `docker` commands to do the setup.

Given that the `bin/setup` script as provided by Rails is useful in its own right, I don't feel comfortable co-opting it.  Especially if we're going to supplant it with `rake` tasks at a future date.

**Comment by bvtsang** [comment]
Reverted changes to `bin/setup` and updated the README.

**Comment by 0xazure** [comment]
So it looks like it was a good thing I waited until now to give things a try, because I'm able to offer a "fresh eyes" perspective.  Unfortunately, I found the REAME to be a bit lacking in details on how to get certain things set up.

First of all, "Build Instructions" isn't exactly accurate.  Should probably be "Setup Instructions".  To wit, I would propose the following improvements to this section:

---

## Setup Instructions

Build the application Docker image:

```sh
$ docker-compose build
```

Start the fleet of containers that support the application:

```sh
$ docker-compose up -d
```

> **Note**: the `-d` flag runs the command in a detached state.  You may have to wait while Docker finishes spinning up the containers before the application or other services are useable.

Create the databases required by Rails:

```sh
$ docker-compose run app rake db:create
```

Get the Linux VM `ip` of Docker's `default` VM:

```sh
docker-machine ip default
```

Vist `<docker-machine ip default>:3000` to confirm the webserver started as expected.

> **Note**: `docker-machine` commands apply only to Windows and Mac OS X (see why [for Mac OS X](https://docs.docker.com/engine/installation/mac/#learn-the-key-concepts-before-installing), [for Windows](https://docs.docker.com/engine/installation/windows/#learn-the-key-concepts-before-installing)).  Linux users can simply visit `localhost:3000` to access the webserver.

---

The section is more verbose now, but this is to the benefit of new dev onboarding.  As I was working though getting set up, even though we'd had conversations about the known issues with the previous `bin/setup`, I _still_ got tripped up on the fact that `-d` makes things run detached, and I was wondering my `docker-compose run -d app rake db:create && rake db:migrate` wasn't working (it was because the PG container hadn't finished starting up yet).

You'll also notice that I removed the `-d` flag from `docker-compose run` command.  Running these commands in a detached state silently swallows any messages the app may return to you, so this is not advisable.

I added a section about `docker-machine` as, since I'm not running on a Linux distro, Docker has to run inside a VM in order to make use of Linux containerization.  This means that addresses like the one the webserver is listening on are different for any dev running the Docker daemon inside a VM, as the IP addresses don't (necessarily) match those on the guest OS.  [This piece](https://docs.docker.com/engine/installation/mac/#access-container-ports) in the documentation was particularly enlightening.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 045abdc90fbdceff0157ce9996b6c743875d7fae
  * Your subject line is longer than 50 characters

Guidelines are available at https://github.com/new-xkit/pipeline/blob/master/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

**Comment by 0xazure** [comment]
LGTM.  Massive :+1: on this, @bvtsang.
